### PR TITLE
[Gutenberg] Add error boundary components and exception logging

### PR DESCRIPTION
**Pull request by fluiddot**
**Related PRs:**
* https://github.com/WordPress/gutenberg/pull/59221
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6655
* https://github.com/Automattic/Automattic-Tracks-iOS/pull/278

This PR enables the error boundary at the editor and block level. It also connects Gutenberg Mobile to the Crash logging service in order to log JavaScript exceptions.

Additionally, it uploads to Sentry the JavaScript source map included in the GBM XCFramework.

**To test:**
Follow testing instructions from https://github.com/wordpress-mobile/gutenberg-mobile/pull/6655.

## Regression Notes
1. Potential unintended areas of impact
None on the user-facing side as this PR mainly implements a new RN bridge function. However, it also adds a step in the build process that if fails might disrupt the release process.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I manually tested the build process for alpha builds. I didn't cover beta and app store builds, but the only error I foresee that it might introduce is tagging the source map files with a wrong release version. We'll work with Apps Infra team closely in the release where these changes are introduced to validate the source map files are properly uploaded.

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: fastlane/lanes/build.rb [hunk 1]
```diff
@@ -175,6 +175,13 @@
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
 
+    upload_gutenberg_sourcemaps(
+      sentry_project_slug: SENTRY_PROJECT_SLUG_WORDPRESS,
+      release_version: release_version_current,
+      build_version: build_code_current,
+      app_identifier: WORDPRESS_BUNDLE_IDENTIFIER
+    )
```
Comment (fluiddot) [inline fastlane/lanes/build.rb:177]: This source map will be used for the App Store builds (beta and production) of the **WordPress app**.

File: fastlane/lanes/build.rb [hunk 2]
```diff
@@ -220,6 +227,13 @@
       project_slug: SENTRY_PROJECT_SLUG_JETPACK,
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
+
+    upload_gutenberg_sourcemaps(
+      sentry_project_slug: SENTRY_PROJECT_SLUG_JETPACK,
+      release_version: release_version_current,
+      build_version: build_code_current,
+      app_identifier: JETPACK_BUNDLE_IDENTIFIER
+    )
```
Comment (fluiddot) [inline fastlane/lanes/build.rb:233]: This source map will be used for the App Store builds (beta and production) of the **Jetpack app**.

File: fastlane/lanes/build.rb [hunk 3]
```diff
@@ -271,6 +285,14 @@
       project_slug: SENTRY_PROJECT_SLUG_WORDPRESS,
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
+
+    upload_gutenberg_sourcemaps(
+      sentry_project_slug: SENTRY_PROJECT_SLUG_WORDPRESS,
+      release_version: release_version_current_internal,
+      build_version: build_code_current_internal,
+      app_identifier: 'org.wordpress.internal'
+    )
```
Comment (fluiddot) [inline fastlane/lanes/build.rb:289]: This source map will be used for the internal builds of the WordPress app. As far as I checked, this version is seems no longer used.

File: fastlane/lanes/build.rb [hunk 4]
```diff
@@ -394,6 +418,13 @@ def build_and_upload_prototype_build(scheme:, output_app_name:, appcenter_app_na
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
 
+    upload_gutenberg_sourcemaps(
+      sentry_project_slug: sentry_project_slug,
+      release_version: release_version_current,
+      build_version: build_number,
+      app_identifier: app_identifier
+    )
```
Comment (fluiddot) [inline fastlane/lanes/build.rb]: This source map will be used for the alpha builds (i.e. installable builds in PRs). The app is determined by the `app_identifier` parameter.

File: fastlane/lanes/build.rb [hunk 5]
```diff
@@ -450,4 +481,41 @@ def upload_build_to_testflight(whats_new_path:)
       changelog: File.read(whats_new_path)
     )
   end
+
+  def upload_gutenberg_sourcemaps(sentry_project_slug:, release_version:, build_version:, app_identifier:)
+    # The bundle and source map files are the same for all architectures.
+    gutenberg_bundle = File.join(PROJECT_ROOT_FOLDER, "Pods/Gutenberg/Frameworks/Gutenberg.xcframework/ios-arm64/Gutenberg.framework")
```
Comment (fluiddot) [inline fastlane/lanes/build.rb]: As specified in the inline comment, the bundle and source map files are included for each architecture. However, both files are the same, as the JavaScript code isn't affected by the architecture. Hence, we can use the files from any of the architectures.

File: Scripts/BuildPhases/CopyGutenbergJS.sh [hunk 6]
```diff
@@ -13,6 +13,7 @@ DEST="$CONFIGURATION_BUILD_DIR/$UNLOCALIZED_RESOURCES_FOLDER_PATH"
 # Update the matching .inputs.xcfilelist when changing these
 XCFRAMEWORK_BUNDLE_ROOT="$PODS_XCFRAMEWORKS_BUILD_DIR/Gutenberg/Gutenberg.framework"
 PODS_BUNDLE_ROOT="$PODS_ROOT/Gutenberg/bundle/ios"
+LOCAL_BUNDLE="$PODS_ROOT/../../gutenberg-mobile/bundle/ios"
```
Comment (fluiddot) [inline Scripts/BuildPhases/CopyGutenbergJS.sh:16]: When testing a production version of the JS bundle in local builds, I noticed we don't have a quick way to use a local bundle. For this reason, I added a third option to fetch the bundle by pointing to a local path.

Comment (fluiddot) [reply Scripts/BuildPhases/CopyGutenbergJS.sh:16]: The option to fetch the bundle form `$PODS_ROOT/Gutenberg/bundle/ios` seems to be deprecated as the library is provided with the XCFramework. We could remove it.

File: fastlane/lanes/build.rb [hunk 7]
```diff
@@ -266,6 +280,14 @@
...
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
+
+    upload_gutenberg_sourcemaps(
+      sentry_project_slug: SENTRY_PROJECT_SLUG_WORDPRESS,
+      release_version: release_version_current_internal,
+      build_version: build_code_current_internal,
+      app_identifier: 'org.wordpress.internal'
+    )
+
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L290"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 8]
```diff
@@ -334,7 +358,7 @@ def generate_prototype_build_number
   # Builds a Prototype Build for WordPress or Jetpack, then uploads it to App Center and comment with a link to it on the PR.
   #
   # rubocop:disable Metrics/AbcSize
-  def build_and_upload_prototype_build(scheme:, output_app_name:, appcenter_app_name:, app_icon:, sentry_project_slug:)
+  def build_and_upload_prototype_build(scheme:, output_app_name:, appcenter_app_name:, app_icon:, sentry_project_slug:, app_identifier:)
```
Comment (dangermattic) [inline fastlane/lanes/build.rb:361]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/a1a692f09379dc1e081781d78ae488238eb956a6/fastlane/lanes/build.rb#L361"></span>Metrics/ParameterLists: Avoid parameter lists longer than 5 parameters. [6/5]</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 9]
```diff
@@ -385,6 +409,13 @@ def build_and_upload_prototype_build(scheme:, output_app_name:, appcenter_app_na
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
 
+    upload_gutenberg_sourcemaps(
+      sentry_project_slug: sentry_project_slug,
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L413"></span>Style/HashSyntax: Omit the hash value.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 10]
```diff
@@ -385,6 +409,13 @@ def build_and_upload_prototype_build(scheme:, output_app_name:, appcenter_app_na
...
+      release_version: release_version_current,
+      build_version: build_number,
+      app_identifier: app_identifier
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L416"></span>Style/HashSyntax: Omit the hash value.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 11]
```diff
@@ -464,4 +495,41 @@ def upload_build_to_app_center(
       notify_testers: false
     )
   end
+
+  def upload_gutenberg_sourcemaps(sentry_project_slug:, release_version:, build_version:, app_identifier:)
+    # The bundle and source map files are the same for all architectures.
+    gutenberg_bundle = File.join(PROJECT_ROOT_FOLDER, "Pods/Gutenberg/Frameworks/Gutenberg.xcframework/ios-arm64/Gutenberg.framework")
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L501"></span>Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 12]
```diff
@@ -464,4 +495,41 @@ def upload_build_to_app_center(
...
+
+    Dir.mktmpdir do |sourcemaps_folder|
+      # It's important that the bundle and source map files have specific names, otherwise, Sentry
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L504"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 13]
```diff
@@ -464,4 +495,41 @@ def upload_build_to_app_center(
...
+      FileUtils.cp(File.join(gutenberg_bundle, 'App.composed.js.map'), File.join(sourcemaps_folder, 'main.jsbundle.map'))
+
+      # To generate the full release version string to attach the source maps, we need to specify:
+      # - App identifier
+      # - Release version
+      # - Build version
+      # This conforms to the following format: <app_identifier>@<release_version>+<build_version>
+      # Here are a couple of examples:
+      # - Prototype build: com.jetpack.alpha@24.2+pr22654-07765b3 
+      # - App Store build: org.wordpress@24.1+24.1.0.3
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L516"></span>Layout/LeadingCommentSpace: Missing space after <code>#</code>.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 14]
```diff
@@ -464,4 +495,41 @@ def upload_build_to_app_center(
...
+
+      sentry_upload_sourcemap(
+        auth_token: get_required_env('SENTRY_AUTH_TOKEN'),
+        org_slug: SENTRY_ORG_SLUG,
+        project_slug: sentry_project_slug,
+        version: release_version,
+        dist: build_version,
+        build: build_version,
+        app_identifier: app_identifier,
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L525"></span>Style/HashSyntax: Omit the hash value.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 15]
```diff
@@ -464,4 +495,41 @@ def upload_build_to_app_center(
...
+        # When the React native bundle is generated, the source map file references
+        # include the local machine path, with the `rewrite` and `strip_common_prefix`
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/f2791d67c867c24143a5e17c6fd2b8f3d5fe55c6/fastlane/lanes/build.rb#L527"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Podfile reference(s) to a branch:<br />
<code>Automattic-Tracks-iOS (from `https://github.com/Automattic/Automattic-Tracks-iOS.git`, branch `add/sentry-js-exception`)</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: fastlane/lanes/build.rb [hunk 16]
```diff
@@ -266,6 +280,14 @@
...
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
+
+    upload_gutenberg_sourcemaps(
+      sentry_project_slug: SENTRY_PROJECT_SLUG_WORDPRESS,
+      release_version: release_version_current_internal,
+      build_version: build_code_current_internal,
+      app_identifier: 'org.wordpress.internal'
+    )
+
```
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/3f5b7cb46a4929942fc8a45bb49d38cf7dc7f8dd/fastlane/lanes/build.rb#L290"></span>Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body end.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb [hunk 17]
```diff
@@ -334,7 +358,8 @@ def generate_prototype_build_number
   # Builds a Prototype Build for WordPress or Jetpack, then uploads it to App Center and comment with a link to it on the PR.
   #
   # rubocop:disable Metrics/AbcSize
-  def build_and_upload_prototype_build(scheme:, output_app_name:, appcenter_app_name:, app_icon:, sentry_project_slug:)
+  # rubocop:disable Metrics/ParameterLists
```
Comment (dangermattic) [inline fastlane/lanes/build.rb:361]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/3f5b7cb46a4929942fc8a45bb49d38cf7dc7f8dd/fastlane/lanes/build.rb#L361"></span>Lint/MissingCopEnableDirective: Re-enable Metrics/ParameterLists cop with <code># rubocop:enable</code> after disabling it.</td>
    </tr>
  </tbody>
</table>

**Review by jhnstn — APPROVED** [review]
The test build worked from https://github.com/wordpress-mobile/gutenberg-mobile/pull/6654) worked as expected.

Looks like there are some linting issues but otherwise looks good to go.
Nice work @fluiddot !

**Comment by fluiddot** [comment]
> Looks like there are some linting issues but otherwise looks good to go.

The only remaining `dangermattic` error is the Tracks reference to a branch. Once https://github.com/Automattic/Automattic-Tracks-iOS/pull/280 is merged, I'll update it with the release tag.

**Pull request merged**